_posts/2021-02-08-ill-be-back.md

```diff
@@ -0,0 +1,60 @@
+
+If we look at time from the perspective of systems design, it has many nuances and complexities.
+On the one hand, there is clock drift: different machines have different views of time; it's not unique and ubiquitous. On the other hand, there is the conceptual complexity of modeling the passage of time. Modeling the passage of time as a clock does seems to be a natural thing to do. It translates well the [user mental model]() into the system design and provides a reliable solution. However, if we look deeper, it might not be so good as a solution.
+
+## Clocks are batch jobs
+
+Time ticks, human beings observe the ticking of time and react.
+
+Imagining a clock scale of 1 day, in which clocks will only tell which day of the week is today. We will be in a situation where there is a long list of things to do today and someone that picks from the list and executes the corresponding task. It'll be a big mess, 24 hours are quite a long time, and those to-do lists can become quite long. It's very tempting to increase the scale and make it so that a clock ticks every one hour instead of every 24. Is anything changed? Might be, we might be lucky and discover that the granularity of a one-hour scale in our context is better. Regardless of the clock scale of choice, we're approaching the problem with a batch job oriented solution. We'll be sitting there waiting for a clock tick to kick off an allocated slot to perform some work regardless of the amount of work.
+
+Let's say that for the amount of work that we need to do during office hours, a 1-hour scale seems a good fit. Every 1 hour, nine to five, "the clock" kicks off a job. The goal is to get done most, if not all, pending charges by the end of the week. The flip side of the coin is that, weekly, the "do xyz" job is kicked off 64 times for no good reasons, during non-office hours, weekends included. A huge waste. One could argue that a simple solution is to turn off the clock when offices are closed. However, this needs another ticking clock.
+
+To better grasp the problems that a batch job oriented solution carries, we can translate the above conceptual view into a more real-world sample by describing an invoicing process.
+
+Acme Inc. issues invoices on the last day of the month. All issued invoices are due in 30 days. It seems reasonable to set the "check overdue invoices" clock to tick on the 10 of the next month. Forty days after sending invoices, Acme checks payments and act upon overdue invoices. One day Acme changed its business model and decided to issue invoices on the same day they sell goods or services. The Acme accounting department cannot anymore rely on the Forty-day rule to check for overdue invoices because by issuing invoices not regularly, invoices can be outstanding more or less every day. The Acme Operations team's simple reaction is to adjust the ticking clock's scale, and instead of ticking on the 10 of the next month, it makes it tick every day.
```

Comment:
```suggestion
Acme Inc. issues invoices on the last day of the month. All issued invoices are due in thirty days. It seems reasonable to set the "check overdue invoices" clock to tick on the 10th of the next month. Forty days after sending invoices, Acme checks payments and acts upon overdue invoices. One day Acme changes its business model and decides to issue invoices on the same day they sell goods or services. The Acme accounting department cannot anymore rely on the forty-day rule to check for overdue invoices because by not issuing invoices regularly, invoices can be outstanding more or less every day. The Acme Operations team's simple reaction is to adjust the ticking clock's scale, and instead of ticking on the 10th of the next month, it makes it tick every day.
```